pkg/controller/machinedeployment/controller.go

```diff
@@ -180,6 +181,40 @@ func (r *ReconcileMachineDeployment) Reconcile(request reconcile.Request) (recon
 		return reconcile.Result{}, errors.Errorf("failed validation on MachineDeployment %q label selector, cannot match any machines ", d.Name)
 	}
 
+	// Cluster might be nil as some providers might not require a cluster object
+	// for machine management.
+	cluster, err := r.getCluster(d)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	// Set the ownerRef with foreground deletion if there is a linked cluster.
+	if cluster != nil && len(d.OwnerReferences) == 0 {
+		blockOwnerDeletion := true
+		d.OwnerReferences = append(d.OwnerReferences, metav1.OwnerReference{
+			APIVersion:         cluster.APIVersion,
+			Kind:               cluster.Kind,
+			Name:               cluster.Name,
+			UID:                cluster.UID,
+			BlockOwnerDeletion: &blockOwnerDeletion,
+		})
+	}
```

Comment:
Should this also return early since updating the OwnerReferences will also lead to the same issues as when we update the Finalizers?

Reply:
I thought the same, but removed the early return given that the when this is set the Finalizer should also be set and return there.